sites/kit.svelte.dev/src/lib/search/SearchBox.svelte

```diff
@@ -0,0 +1,416 @@
+<script>
+	import { onMount } from 'svelte';
+	import { Icon } from '@sveltejs/site-kit';
+	import flexsearch from 'flexsearch';
+	import { afterNavigate } from '$app/navigation';
+	import { searching, query, recent } from './stores.js';
+	import { focusable_children, trap } from '../actions/focus.js';
+
+	let modal;
+
+	let results = [];
+	let backspace_pressed;
+
+	let index;
+	let lookup;
+
+	onMount(async () => {
+		const response = await fetch('/content.json');
+		const { blocks } = await response.json();
+
+		index = new flexsearch.Index({
+			tokenize: 'forward'
+		});
+
+		lookup = new Map();
+
+		let time = Date.now();
+		for (const block of blocks) {
+			const title = block.breadcrumbs[block.breadcrumbs.length - 1];
+			lookup.set(block.href, {
+				title,
+				href: block.href,
+				breadcrumbs: block.breadcrumbs.slice(0, -1),
+				content: block.content
+			});
+			index.add(block.href, `${title} ${block.content}`);
+
+			// poor man's way of preventing blocking
+			if (Date.now() - time > 25) {
+				await new Promise((f) => setTimeout(f, 0));
+				time = Date.now();
+			}
+		}
+
+		update();
+	});
+
+	afterNavigate(() => {
+		// TODO this also needs to apply when only the hash changes
+		// (should before/afterNavigate fire at that time? unclear)
+		$searching = false;
+	});
+
+	function update() {
+		results = (index ? index.search($query) : []).map((href) => lookup.get(href));
+	}
+
+	function escape(text) {
+		return text.replace(/</g, '&lt;').replace(/>/g, '&gt;');
+	}
+
+	function excerpt(content, query) {
+		const index = content.toLowerCase().indexOf(query.toLowerCase());
+		if (index === -1) {
+			return content.slice(0, 100);
+		}
+
+		const prefix = index > 20 ? `…${content.slice(index - 15, index)}` : content.slice(0, index);
+		const suffix = content.slice(
+			index + query.length,
+			index + query.length + (80 - (prefix.length + query.length))
+		);
+
+		return (
+			escape(prefix) +
+			`<mark>${escape(content.slice(index, index + query.length))}</mark>` +
+			escape(suffix)
+		);
+	}
+
+	/** @param {string} href */
+	function navigate(href) {
+		$recent = [href, ...$recent.filter((x) => x !== href)];
+		$searching = false;
+	}
+
+	$: if ($searching) update();
+
+	$: recent_searches = lookup ? $recent.map((href) => lookup.get(href)).filter(Boolean) : [];
+</script>
+
+<svelte:window
+	on:keydown={(e) => {
+		if (e.code === 'KeyK' && (navigator.platform === 'MacIntel' ? e.metaKey : e.ctrlKey)) {
+			e.preventDefault();
+			$query = '';
+			$searching = !$searching;
+		}
+
+		if (e.code === 'Escape') {
+			$searching = false;
+		}
+	}}
+/>
+
+{#if $searching && index}
+	<div
+		class="modal-background"
+		on:click={() => ($searching = false)}
+		on:wheel={(e) => e.preventDefault()}
```

Comment:
Nit: when I hover over the modal and scroll, the underlying page still scrolls. This only happens if the modal is not showing a scrollbar (i.e. not enough results to scroll). Not sure if it's worth addressing.